web_monitoring/html_diff_render.py

```diff
@@ -179,30 +364,81 @@ def _diff_title(old, new):
-    the "new" element with its content replaced by the diff.
-    """
+def diff_elements(old, new, include='all'):
+    results = {}
     if not old or not new:
-        return ''
-    result_element = copy.copy(new)
-    result_element.clear()
-    result_element.append(_htmldiff(str(old), str(new)))
-    return result_element
+        return results
+
+    def fill_element(element, diff):
+        result_element = copy.copy(element)
+        result_element.clear()
+        result_element.append(diff)
+        return result_element
 
+    metadata, raw_diffs = _htmldiff(str(old), str(new), include)
+    for diff_type, diff in raw_diffs.items():
+        element = diff_type == 'deletions' and old or new
+        results[diff_type] = fill_element(element, diff)
 
-def _htmldiff(old, new):
+    return metadata, results
+
+
+def _htmldiff(old, new, include='all'):
     """
     A slightly customized version of htmldiff that uses different tokens.
     """
     old_tokens = tokenize(old)
     new_tokens = tokenize(new)
-    old_tokens = [_customize_token(token) for token in old_tokens]
-    new_tokens = [_customize_token(token) for token in new_tokens]
-    result = htmldiff_tokens(old_tokens, new_tokens)
-    result = ''.join(result).strip()
-    return fixup_ins_del_tags(result)
+    # old_tokens = [_customize_token(token) for token in old_tokens]
+    # new_tokens = [_customize_token(token) for token in new_tokens]
+    old_tokens = _customize_tokens(old_tokens)
+    new_tokens = _customize_tokens(new_tokens)
+    # result = htmldiff_tokens(old_tokens, new_tokens)
+    # result = diff_tokens(old_tokens, new_tokens) #, include='delete')
+    logger.debug('CUSTOMIZED!')
+
+    # HACK: The whole "spacer" token thing above in this code triggers the
+    # `autojunk` mechanism in SequenceMatcher, so we need to explicitly turn
+    # that off. That's probably not great, but I don't have a better approach.
+    matcher = InsensitiveSequenceMatcher(a=old_tokens, b=new_tokens, autojunk=False)
+    # matcher = SequenceMatcher(a=old_tokens, b=new_tokens, autojunk=False)
+    opcodes = matcher.get_opcodes()
+
+    metadata = _count_changes(opcodes)
+    diffs = {}
+
+    def render_diff(diff_type):
+        diff = assemble_diff(old_tokens, new_tokens, opcodes, diff_type)
+        # return fixup_ins_del_tags(''.join(diff).strip())
+        result = ''.join(diff).strip().replace('</li> ', '</li>')
+        return result
+
+    if include == 'all' or include == 'combined':
+        diffs['combined'] = render_diff('combined')
+    if include == 'all' or include == 'insertions':
+        diffs['insertions'] = render_diff('insertions')
+    if include == 'all' or include == 'deletions':
+        diffs['deletions'] = render_diff('deletions')
+
+    return metadata, diffs
+
+
+def _count_changes(opcodes):
+    results = {
+        'change_count': 0,
+        'deletion_count': 0,
+        'insertion_count': 0,
+    }
+
+    for command, *rest in opcodes:
```

Comment:
`itertools.Counter` could be used to simplify and possibly speed up this process.

Reply:
Hmmm… are you thinking something like:

```py
counts = Counter(map(lambda operation: operation[0], opcodes)
return {
    'change_count': counts['insert'] + counts['replace'],
    'deletion_count': counts['delete'] + counts['replace'],
    'insertion_count': counts['insert'] + counts['delete'] + counts['replace'],
}
```

?

Reply:
Yes. I almost feel guilty for suggesting it because there are so many bigger fish to fry. But yes.

Reply:
No, that’s great. I think I’ve only used `Counter` once before even though it’s pretty neat. This seems like a good use case.